tools/c7n_azure/c7n_azure/resources/cosmos_db.py

```diff
@@ -403,6 +404,158 @@ def _query_rules(self, resource):
+     which in this case means we don't remove any previously configured
+     vnet firewall rules.
+
+     .. code-block:: yaml
+
+        policies:
+          - name: cosmos-firewall
+            resource: azure.cosmosdb
+            filters:
+              # The firewall is disabled
+              - type: value
+                key: properties.ipRangeFilter
+                value: empty
+            actions:
+              - type: set-firewall-rules
+                append: True
+                bypass-rules:
+                  - AzureCloud
+                  - Portal
+                ip-rules:
+                  - 19.0.0.0/16
+                  - 20.0.1.2
+
+
+     Cosmos firewalls are disabled by simply configuring them with empty values.
+     We can do this with an empty action, which defaults to ``append: False``.
+
+     .. code-block:: yaml
+
+        policies:
+          - name: cosmos-firewall-clear
+            resource: azure.cosmosdb
+            filters:
+              # The firewall is enabled
+              - not:
+                - type: value
+                  key: properties.ipRangeFilter
+                  value: empty
+            actions:
+              - type: set-firewall-rules
+
+
+     """
+    schema = type_schema(
+        'set-firewall-rules',
+        required=[],
+        **{
+            'append': {'type': 'boolean', "default": False},
+            'bypass-rules': {'type': 'array', 'items': {
+                'enum': ['Portal', 'AzureCloud']}},
+            'ip-rules': {'type': 'array', 'items': {'type': 'string'}},
+            'virtual-network-rules': {'type': 'array', 'items': {'type': 'string'}}
+        }
+    )
+
+    def __init__(self, data, manager=None):
+        super(CosmosSetFirewallAction, self).__init__(data, manager)
+        self._log = logging.getLogger('custodian.azure.cosmosdb')
+        self.rule_limit = 1000
+
+    def _process_resource(self, resource):
+        existing_ip = filter(None, resource['properties'].get('ipRangeFilter', '').split(','))
+        rules = self._build_ip_rules(existing_ip, self.data.get('ip-rules', []))
+
+        # Cosmos DB does not have real bypass
+        # instead the portal UI adds these values to your
+        # rules filter when you check the box.
+        bypass = self.data.get('bypass-rules', [])
+        if 'Portal' in bypass:
+            rules.extend(['104.42.195.92',
+                          '40.76.54.131',
+                          '52.176.6.30',
+                          '52.169.50.45',
+                          '52.187.184.26'])
+        if 'AzureCloud' in bypass:
+            rules.append('0.0.0.0')
+
+        # If the user has too many rules log and skip
+        if len(rules) > self.rule_limit:
+            self._log.error("Skipped updating firewall for %s. "
+                            "%s exceeds maximum rule count of %s." %
+                            (resource['name'], len(rules), self.rule_limit))
+            return
```

Comment:
Better to raise an exception here.. If you return, we treat it as a success, so when we add metrics for failed actions count they will be incorrect.

Reply:
I think the base action class eats all exceptions and returns anyway.

But regardless we usually just warn on a single resource failing though right? 

I know in this case you could easily be in a situation where every resource is going to be skipped, but if it is an append situation that won't be true.